File: knapsack.c
/*
Adnan Sulaiman
20CO13
*/

#include<stdio.h>
void knapsack();
int main()
{
    int n,cap;
    printf("Enter the number of weights");
    scanf("%d",&n);
    printf("Enter the capacity");
    scanf("%d",&cap);
    float wt[n],prof[n],ratio[n];
    printf("Enter the weights and its profit");
    for(int i=0;i<n;i++)
    {
        scanf("%f %f",&wt[i],&prof[i]);
        ratio[i]=prof[i]/wt[i];
    }
    float temp=0;

    for(int i=0;i<n;i++)
    {
        for(int j=i+1;j<n;j++)
        {
            if(ratio[i]<ratio[j])
            {
                temp=ratio[i];
                ratio[i]=ratio[j];
                ratio[j]=temp;

                temp=wt[i];
                wt[i]=wt[j];
                wt[j]=temp;

                temp=ratio[i];
                prof[i]=prof[j];
                prof[j]=temp;

                
            }
        }
    }
    for(int i=0;i<n;i++)
    {
        printf("%f\n",ratio[i]);
    }


    knapsack(n,wt,prof,cap);

    return 0;
}

void knapsack(int n,float WT[],float PROF[],int cap)
{
    int i;
    float x[n],tp=0,c;
    c=cap;

    for(i=0;i<n;i++)
    {
        if(WT[i]>c)
        break;

        else{
            c=c-WT[i];
            tp=tp+PROF[i];
            x[i]=1;
        }
    }
    if(i<n)
    {
        x[i]=c/WT[i];
        tp=tp+PROF[i]*x[i];
    }
    
    printf("The ratios are \n");
    for(int i=0;i<n;i++)
    {
        printf("%f\n",x[i]);
    }

    printf("The total profit is %f", tp);
    
}

File: knapsack.java
/*
    Name : Khan Arshad Abdulla
    Roll No : 20CO24
*/

import java.util.*;
class knapsack{

    void greedy(int n, double wt[],double pr[],double cap){
        int i;
        double u = cap;
        double total = 0;
        double x[]=new double[n];
        // for(i=0;i<n;i++){lcs using dynamic programming code in c
        //     x[i]=0.0;
        // }
        for(i=0;i<n;i++){
            if(wt[i]>u){
                break;
            }
            else{
                x[i]=1.0;
                total = total + pr[i];
                u = u - wt[i];
            }
        }
        if(i<n){
            x[i] = u/wt[i]; 
            total = total + (x[i]*pr[i]);
        }
        System.out.print("The Resultant Vector: ");
        for(i=0;i<n;i++){
            System.out.print("\t"+x[i]);
        }
        System.out.print("\nTotal Profit: "+total);
    }
    public static void main(String arg[]){
        int i,j,n;
        Scanner sc=new Scanner(System.in);
        System.out.print("Enter No of objects: ");
        n=sc.nextInt();
        double wt[]=new double[n];
        double pr[]=new double[n];
        System.out.print("Enter Weights & Profit: ");
        for(i=0;i<n;i++){
            wt[i]=sc.nextFloat();
            pr[i]=sc.nextFloat();
        }
        System.out.print("Enter Capacity of Sack: ");
        double cap=sc.nextFloat();
        double ratio[]=new double[n];
        for(i=0;i<n;i++){
            ratio[i] = pr[i] / wt[i];
        }

        for(i=0;i<n;i++){
            for(j=i+1;j<n;j++){
                if(ratio[j]>ratio[i]){
                    double temp;
                    temp = ratio[j];
                    ratio[j] = ratio[i];
                    ratio[i]=temp;
                }
                if(pr[j]>pr[i]){
                    double temp = pr[j];
                    pr[j] = pr[i];
                    pr[i]=temp;
                }
                if(wt[j]>wt[i]){
                    double temp = wt[j];
                    wt[j] = wt[i];
                    wt[i]=temp;
                }
            }
        }
        knapsack ob = new knapsack();
        ob.greedy(n, wt, pr, cap);
        // for(i=0;i<n;i++){
        //     System.out.print(wt[i]+"\t");
        // }
        // for(i=0;i<n;i++){
        //     System.out.print("\t"+pr[i]);
        // }
    }
}